File: backend/internal/scheduler/sjf_preemptive.go
package scheduler

import (
    "github.com/pdro1812/trabalho-os-scheduler/backend/internal/models"
)

func RunSJFPreemptive(req models.SimulationRequest) models.SimulationResponse {
    processes := make([]models.Process, len(req.Processes))
    copy(processes, req.Processes)

    n := len(processes)
    remainingTime := make([]int, n)
    for i := range processes {
        remainingTime[i] = processes[i].BurstTime
    }

    var executionOrder []models.ExecutionBlock
    var metrics []models.ProcessMetrics

    currentTime := 0
    completed := 0
    lastPID := ""
    var blockStart int

    totalWaitTime := 0
    totalTurnaroundTime := 0

    for completed < n {
        idx := -1
        minRT := 999999999

        // Procura quem tem o menor tempo restante NESTE EXATO MILISSEGUNDO
        for i := 0; i < n; i++ {
            if processes[i].ArrivalTime <= currentTime && remainingTime[i] > 0 {
                if remainingTime[i] < minRT {
                    minRT = remainingTime[i]
                    idx = i
                } else if remainingTime[i] == minRT && idx != -1 {
                    // Desempate: quem chegou primeiro
                    if processes[i].ArrivalTime < processes[idx].ArrivalTime {
                        idx = i
                    }
                }
            }
        }

        if idx != -1 {
            p := processes[idx]

            // Se a CPU trocou de processo (Preempção ou Fim de execução anterior)
            if lastPID != p.PID {
                // Salva o bloco do processo anterior
                if lastPID != "" && blockStart < currentTime {
                    executionOrder = append(executionOrder, models.ExecutionBlock{
                        PID:   lastPID,
                        Start: blockStart,
                        End:   currentTime,
                    })
                }

                // Aplica o TTC
                if lastPID != "" && req.TTC > 0 {
                    executionOrder = append(executionOrder, models.ExecutionBlock{
                        PID:   "TTC",
                        Start: currentTime,
                        End:   currentTime + req.TTC,
                    })
                    currentTime += req.TTC
                    lastPID = "" // Reseta para forçar a reavaliação (alguém pode ter chegado durante o TTC)
                    continue
                }

                lastPID = p.PID
                blockStart = currentTime
            }

            // Roda por 1 unidade de tempo (Tick)
            remainingTime[idx]--
            currentTime++

            // Se o processo acabou neste tick
            if remainingTime[idx] == 0 {
                completed++
                executionOrder = append(executionOrder, models.ExecutionBlock{
                    PID:   p.PID,
                    Start: blockStart,
                    End:   currentTime,
                })

                turnaround := currentTime - p.ArrivalTime
                wait := turnaround - p.BurstTime
                totalTurnaroundTime += turnaround
                totalWaitTime += wait

                metrics = append(metrics, models.ProcessMetrics{
                    PID:           p.PID,
                    EffectiveTime: turnaround,
                    WaitingTime:   wait,
                })

                // CORREÇÃO: Avançamos o blockStart para evitar salvar esse bloco duplicado na próxima troca, 
                // mas MANTEMOS o lastPID para que o próximo processo acione o TTC.
                blockStart = currentTime
            }
        } else {
            // CPU Ociosa
            // CORREÇÃO: Como a CPU ficou ociosa, limpamos o contexto. O próximo processo não deve sofrer TTC de saída.
            lastPID = ""
            
            nextArrival := 999999999
            for i := 0; i < n; i++ {
                if remainingTime[i] > 0 && processes[i].ArrivalTime < nextArrival {
                    nextArrival = processes[i].ArrivalTime
                }
            }
            currentTime = nextArrival
        }
    }

    return models.SimulationResponse{
        ExecutionOrder:    executionOrder,
        Metrics:           metrics,
        AvgWaitTime:       float64(totalWaitTime) / float64(n),
        AvgTurnaroundTime: float64(totalTurnaroundTime) / float64(n),
    }
}

File: backend/internal/scheduler/sjf_preemptive.js
function runSJFPreemptive(requisicao) {
    const processos = [];
    for (let i = 0; i < requisicao.Processes.length; i++) {
        processos.push(requisicao.Processes[i]);
    }

    const n = processos.length;
    const tempoRestante = [];
    for (let i = 0; i < n; i++) {
        tempoRestante.push(processos[i].BurstTime);
    }

    const ordemExecucao = [];
    const metricas = [];
    let tempoAtual = 0;
    let concluidos = 0;
    let somaEspera = 0;
    let somaTurnaround = 0;

    // Variáveis para controlar o desenho do gráfico de Gantt
    let ultimoPID = "";
    let inicioDoBloco = 0;

    // Flag para saber se a CPU estava parada
    let veioDeOciosidade = true;

    // Loop que roda enquanto houver processos para terminar
    while (concluidos < n) {
        
        // --- PASSO 1: Encontrar o processo com menor tempo restante ---
        let indiceEscolhido = -1;
        let menorTempo = Infinity;

        for (let i = 0; i < n; i++) {
            // O processo precisa ter chegado E ainda ter tempo para rodar
            if (processos[i].ArrivalTime <= tempoAtual && tempoRestante[i] > 0) {
                if (tempoRestante[i] < menorTempo) {
                    menorTempo = tempoRestante[i];
                    indiceEscolhido = i;
                }
                // Em caso de empate, o que chegou primeiro tem prioridade
                else if (tempoRestante[i] === menorTempo) {
                    if (processos[i].ArrivalTime < processos[indiceEscolhido].ArrivalTime) {
                        indiceEscolhido = i;
                    }
                }
            }
        }

        // --- PASSO 2: Se ninguém puder rodar agora, a CPU fica ociosa ---
        if (indiceEscolhido === -1) {
            // Se alguém estava rodando antes, fechamos o bloco dele no Gantt
            if (ultimoPID !== "" && inicioDoBloco < tempoAtual) {
                ordemExecucao.push({ PID: ultimoPID, Start: inicioDoBloco, End: tempoAtual });
            }

            ultimoPID = "";
            veioDeOciosidade = true;

            // Avançamos o tempo para a próxima chegada
            let proximaChegada = Infinity;
            for (let i = 0; i < n; i++) {
                if (tempoRestante[i] > 0 && processos[i].ArrivalTime < proximaChegada) {
                    proximaChegada = processos[i].ArrivalTime;
                }
            }
            tempoAtual = proximaChegada;
            inicioDoBloco = tempoAtual;
            continue;
        }

        const p = processos[indiceEscolhido];

        // --- PASSO 3: Lidar com a troca de processo (Preempção ou Início) ---
        if (p.PID !== ultimoPID) {
            
            // Fecha o bloco do processo anterior no Gantt
            if (ultimoPID !== "" && inicioDoBloco < tempoAtual) {
                ordemExecucao.push({ PID: ultimoPID, Start: inicioDoBloco, End: tempoAtual });
            }

            // Se mudou de processo e não veio de ociosidade, cobra TTC
            if (veioDeOciosidade === false && ultimoPID !== "" && requisicao.TTC > 0) {
                ordemExecucao.push({
                    PID: "TTC",
                    Start: tempoAtual,
                    End: tempoAtual + requisicao.TTC
                });
                tempoAtual = tempoAtual + requisicao.TTC;
                
                // Após o TTC, precisamos reavaliar quem deve rodar (novos processos podem ter chegado)
                inicioDoBloco = tempoAtual;
                ultimoPID = ""; 
                continue; 
            }

            // Começa um novo bloco para o novo processo
            ultimoPID = p.PID;
            inicioDoBloco = tempoAtual;
            veioDeOciosidade = false;
        }

        // --- PASSO 4: Executar 1 unidade de tempo (Tick) ---
        tempoRestante[indiceEscolhido]--;
        tempoAtual++;

        // --- PASSO 5: Se o processo terminou, calculamos as métricas ---
        if (tempoRestante[indiceEscolhido] === 0) {
            concluidos++;

            // Fecha o bloco no Gantt
            ordemExecucao.push({ PID: p.PID, Start: inicioDoBloco, End: tempoAtual });

            const turnaround = tempoAtual - p.ArrivalTime;
            const espera = turnaround - p.BurstTime;

            metricas.push({
                PID: p.PID,
                EffectiveTime: turnaround,
                WaitingTime: espera
            });

            somaTurnaround += turnaround;
            somaEspera += espera;

            // Prepara para o próximo bloco
            inicioDoBloco = tempoAtual;
        }
    }

    return {
        ExecutionOrder: ordemExecucao,
        Metrics: metricas,
        AvgWaitTime: somaEspera / n,
        AvgTurnaroundTime: somaTurnaround / n
    };
}

module.exports = { runSJFPreemptive };